x-pack/plugins/fleet/server/services/artifacts/artifacts.ts

```diff
@@ -84,55 +88,106 @@ export const createArtifact = async (
+  const sortedArtifacts = sortBy(artifacts, 'encodedSize');
 
-  const body = ids.flatMap((id, index) => [
-    {
+  sortedArtifacts.forEach((artifact, index) => {
+    const esArtifactResponse = esSearchHitToArtifact({
+      _id: uniqueIdFromArtifact(artifact),
+      _source: newArtifactToElasticsearchProperties(artifacts[index]),
+    });
+    const esArtifact = newArtifactToElasticsearchProperties(artifact);
+    const bulkOperation = {
       create: {
-        _id: id,
+        _id: uniqueIdFromArtifact(artifact),
       },
-    },
-    newArtifactsData[index],
-  ]);
+    };
 
-  const res = await withPackageSpan('Bulk create fleet artifacts', () =>
-    esClient.bulk({
-      index: FLEET_SERVER_ARTIFACTS_INDEX,
-      body,
-      refresh,
-    })
-  );
-  if (res.errors) {
-    const nonConflictErrors = res.items.reduce<Error[]>((acc, item) => {
-      if (item.create?.status !== 409) {
-        acc.push(new Error(item.create?.error?.reason));
+    // Before adding the next artifact to the current batch, check if it can be added depending on the batch size limit.
+    // If there is no artifact yet added to the current batch, we add it anyway ignoring the batch limit as the batch size has to be > 0.
+    if (artifact.encodedSize + artifactsBatchLengthInBytes >= maxLengthBytes) {
+      artifactsBatchLengthInBytes = artifact.encodedSize;
+      // Use non sorted artifacts array to preserve the artifacts order in the response
+      artifactsEsResponse.push(esArtifactResponse);
+
+      batches.push([bulkOperation, esArtifact]);
+    } else {
+      // Case it's the first one
+      if (isEmpty(batches)) {
+        batches.push([]);
       }
-      return acc;
-    }, []);
-    if (nonConflictErrors.length > 0) {
-      return { errors: nonConflictErrors };
+      // Adds the next artifact to the current batch and increases the batch size count with the artifact size.
+      artifactsBatchLengthInBytes += artifact.encodedSize;
+      // Use non sorted artifacts array to preserve the artifacts order in the response
+      artifactsEsResponse.push(esArtifactResponse);
+
+      batches[batches.length - 1].push(bulkOperation, esArtifact);
     }
+  });
+
+  return { batches, artifactsEsResponse };
+};
+
+export const bulkCreateArtifacts = async (
+  esClient: ElasticsearchClient,
+  artifacts: NewArtifact[],
+  refresh = false,
+  createArtifactsBulkBatchSize: number = MAX_LENGTH_BYTES
+): Promise<{ artifacts?: Artifact[]; errors?: Error[] }> => {
+  const { batches, artifactsEsResponse } = generateArtifactBatches(
+    artifacts,
+    createArtifactsBulkBatchSize
+  );
+
+  const logger = appContextService.getLogger();
+  const nonConflictErrors = [];
+  logger.debug(`Number of batches generated for fleet artifacts: ${batches.length}`);
+  for (let batchN = 0; batchN < batches.length; batchN++) {
+    logger.debug(
+      `Creating artifacts for batch ${batchN + 1} with ${batches[batchN].length / 2} artifacts`
```

Comment:
Q: This may produce a floating number message no? For instance, `Creating artifacts for batch 2 with 3.5 artifacts`? Half an artifact doesn't make sense. Is the length of the batch always an even number?

Reply:
Yes, it is an even number as the body has a pair of values of bulk create action object and the artifact itself

Reply:
I see. So `batches[batchN].length` is always some even number?

Reply:
Correct: `[{bulkAction}, {Artifact}, {bulkAction}, {Artifact}...]`
See this line: https://github.com/elastic/kibana/pull/159187/files#diff-231337d0ab0963e2654e75eecdbdabf8a771c3e3be716a5a93333bf972a4f272R137